FLOE-207: Added tooltips

jobara opened the pull request:
Implemented the tooltip feature for bindDialog and updated the styles to match the latest wireframes.

http://issues.fluidproject.org/browse/FLOE-207

cindyli committed:
FLOE-207: Added tooltips for feedback buttons.

jobara committed:
FLOE-207: Some styling clean up

Cleaned up the styling for the arrow a bit, and added styling to prevent the icons from being underlined. This wasn't noticeable because foundation took care of it, but the feedback tool needs to work without foundation present.

jobara committed:
FLOE-207: Updated the colours

Changed the colour of the arrow and the dialog border to match the updated wireframes.

jobara committed:
FLOE-207: Added styles for the tooltip

As part of this change refactored the css into a single file, and moved the gpii-feedback-button class to the <li> from the <a>.

jobara committed:
FLOE-207: Added the bar hover style

Added the bar hover style that goes over the icons.

jobara committed:
FLOE-207: Removed the icon colour inversion

Removed the colour inversion hover styling from the icons.

jobara committed:
FLOE-207: Added the arrow focus styling

On focusin and focusout a class is added/removed from the bindDialog container.

jobara committed:
Merge branch 'master' into FLOE-207

jobara committed:
FLOE-207: Added focus styling.

jobara committed:
FLOE-207: Update theme styling for tooltip

jobara committed:
Merge remote-tracking branch 'cindy/FLOE-194' into FLOE-207

Conflicts:
	src/components/feedback/js/bindDialog.js

jobara committed:
FLOE-207: Adjusted when the tooltip is shown

The tooltip will be removed when the dialog is open, and reappear after the dialog has closed. The ESC key had to be unbound from closing the dialog to preven 1) styling issues 2) from being closed when the dialog was closed with the ESC key.

jobara committed:
Merge branch 'master' into FLOE-207

jobara committed:
FLOE-207: Refactored unit tests and fixed bugs

Refactored and expanded bindDialog unit tests using the IoCTesting framework. Also fixed some bugs exposed by the tests.

jobara committed:
FLOE-207: Added tooltip test

Added a test to ensure that the tooltip component is created. Also added in the jquery.simulate.js plugin which should have been added to the the previous commit.

jobara committed:
FLOE-207: hover and focus on icon

Moved the hover styling and focusing down to the icon instead of the <li> that is used as the container. This removes some oddities with the tooltip where both the container and the icon could cause it to appear.

jobara committed:
FLOE-207: Arrow only shows with dialog and tooltip

The arrow indicatory is now only shown when the dialog or the tooltip is open. The adding/removal of this class is handled by the model values for isDialogOpen and isTooltipOpen.

jobara committed:
FLOE-207: Added unit tests for indicator state

jobara committed:
FLOE-207: Added font info

jobara committed:
FLOE-207: removed README.md

The text version renders fine in github and it is easier to edit and read in an editor.

amb26 commented on src/components/feedback/js/bindDialog.js:
```diff
@@ -92,20 +142,72 @@ var gpii = gpii || {};
                 method: "dialog",
                 args: "open"
+            },
+            "onFocusIcon.addFocusClass": {
```
Please could you replace this repeated configuration and binding with a simple declarative piece of micro-framework similar to the one in https://github.com/fluid-project/videoPlayer/blob/master/js/VideoPlayer_showHide.js (see FLUID-4928 for some discussion)

jobara committed:
FLOE-207: Corrected namespace

jobara committed:
FLOE-207: Refactored focus and hover binding

Moved the focus and hover to model values and listen to model changes instead.

jobara commented:
@amb26 I've refactored the focus/blur and hover listeners to use model values and bind to modelListeners.

amb26 commented on src/components/feedback/js/bindDialog.js:
```diff
@@ -101,11 +193,33 @@ var gpii = gpii || {};
+                "gpii.metadata.feedback.handleIndicatorState({that}.container, {that}.model, {that}.options.styles.openIndicator)"
+            ],
+            "isTooltipOpen": "gpii.metadata.feedback.handleIndicatorState({that}.container, {that}.model, {that}.options.styles.openIndicator)",
+            "isFocused.icon": {
```
Hi @jobara - could you unwrap one more level of declarative configuration, and subsume these handlers into an area of configuration called "modelToClass" or some such (naturally operated by a suitable grade) - which, for example, takes configuration of the form: 
"modelToClass": {
    "isFocused.icon": "focus",
    "isHovered.icon": "hover"
}

This should let us eliminate the last redundancy in our listeners - cheers

jobara committed:
FLOE-207: Put modelListener setup in modelToClass

Added a new grade that looks through a modelToClass config block to generated model listeners for toggling classes on the component's container.

jobara commented:
@amb26 ready for another round of review

acheetham commented:
I just checked with Joanna, and she said that the buttons on the left of the toolbar also need tooltips. The text is specified on the wiki as:

Preferences: Name of preference is/are status   "Transcripts are available" "Transcripts are available for 1/2 videos" "Transcripts are not available"

http://wiki.fluidproject.org/display/fluid/Metadata+Feedback

amb26 commented on src/components/feedback/js/bindDialog.js:
```diff
@@ -101,11 +212,27 @@ var gpii = gpii || {};
+                icon: false
+            }
+        },
+        modelToClass: {
```
Looks great now - do add a comment here indicating which grade is responsible for handling this config. This is a great sketch for the "new renderer", and we should watch out for other components we can use it in.

amb26 commented on src/components/feedback/js/bindDialog.js:
```diff
@@ -41,18 +98,64 @@ var gpii = gpii || {};
                 }
+            },
+            tooltip: {
+                type: "fluid.tooltip",
```
Consider factoring out all of this configuration into a separate grade, so that it is more reusable and the parent definition is easier to use. What you could do is create two grades - 
gpii.metadata.feedback.dialogTooltip - holding this definition - but with the name {bindDialog} replaced by, say, {tooltipHolder}. 
and
gpii.metadata.feedback.tooltipHolder - which is empty, but applied to "bindDialog" so that it can be IoC resolved. The only piece of non-generic looking configuration is {bindDialog}.options.strings.buttonLabel - which you can leave behind as the sole piece of direct configuration which remains here inline. e.g. this definition would now appear as
tooltip: {
    type: "gpii.metadata.feedback.dialogTooltip",
    container: "{bindDialog}.dom.icon",
    options: {
       content: "{bindDialog}.options.strings.buttonLabel"
    }

and everything else would go into the grade definition.

amb26 commented on src/components/feedback/js/bindDialog.js:
```diff
@@ -74,7 +177,11 @@ var gpii = gpii || {};
             onBindDialogHandlers: null,
-            onDialogReady: null
+            onDialogReady: null,
+            onFocusIcon: null,
```
I believe these events are now redundant and can be removed - the user can now subscribe to the relevant change events from the model fields

amb26 commented on src/css/fonts/selection.json:
```diff
@@ -0,0 +1,539 @@
+{
+	"IcoMoonType": "selection",
```
This file should have tabs normalised to spaces

amb26 commented on src/components/feedback/js/bindDialog.js:
```diff
@@ -211,4 +350,22 @@ var gpii = gpii || {};
+    // which should trigger the tooltip to reappear.
+    gpii.metadata.feedback.reopenTooltip = function (tooltip, isDialogOpen) {
+        if (!isDialogOpen) {
+            tooltip.container.tooltip("open");
```
Could we use the tooltip component's public (that is, Infusion) API here? If not, we should file it as a bug against our component.

amb26 commented on src/components/feedback/js/bindDialog.js:
```diff
@@ -41,18 +98,64 @@ var gpii = gpii || {};
+                            args: ["{that}.container"]
+                        },
+                        "afterOpen.updateModel": {
+                            func: "{bindDialog}.applier.change",
```
more idiomatic here is:
"afterOpen.updateModel": {
    changePath: "{bindDialog}.model.isTooltipOpen",
    value: true
    }
see https://github.com/fluid-project/infusion-docs/blob/master/ChangeApplierAPI.md#firing-a-change-using-a-changeapplier

jobara commented:
@acheetham the tooltips for the metadata indicator icons will be added as part of http://issues.fluidproject.org/browse/FLOE-223

jobara committed:
Merge branch 'master' into FLOE-207

Conflicts:
	demos/feedback/index.html
	tests/all-tests.html
	tests/components/feedback/html/bindDialog-Test.html
	tests/components/feedback/js/bindDialogTests.js

jobara committed:
FLOE-207: Added a comment for modelToClass option

Added a comment describing where the functionality for the modelToClass option is bound.

jobara committed:
FLOE-207: Removed unused events.

Missed removing old events after refactoring the hover and focus states into the model.

jobara committed:
FLOE-207: Converting tabs to spaces

jobara committed:
FLOE-207: Using tooltip wrapper api

jobara committed:
FLOE-207: Improved model change declaration

Using the more idiomatic style of changing model values from a listener.

jobara committed:
FLOE-207: Refactored the tooltip subcomponent

Refactored the tooltip subcomponent into a separate grade.

jobara commented:
@amb26 ready for another round of review

amb26 commented on src/components/feedback/js/bindDialog.js:
```diff
@@ -31,8 +31,115 @@ var gpii = gpii || {};
+
+    /*
+     * This is stricktly to be used as a subcomponent.
+     * Expects the "gpii.metadata.feedback.tooltipHolder" grade to be supplied
```
Should reword this comment requirement to be a little weaker - what the grade relies on is that
i) There is a parent component resolvable with the context {tooltipHolder}
ii) It has a model with a field "isTooltipOpen" with boolean value
iii) It is a ViewComponent

jobara committed:
FLOE-207: Updated a comment

Updated the comment for the gpii.metadata.feedback.dialogTooltip grade.

jobara commented:
@amb26 I've updated the comment
